src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py

```diff
@@ -415,91 +402,57 @@ def loadFiles(self):
+
+                    currentBlock = self.getBlock(files[0], location, True)
+                    currentBlock.setAcquisitionEra(era = dspInfo['AcquisitionEra'])
+                    currentBlock.setProcessingVer(procVer = dspInfo['ProcessingVer'])
+
+                    for newFile in files:
+
+                        if not newFile.get('block', 1) == None:
+                            # Then this file already has a block
+                            # It should be accounted for somewhere
+                            # Or loaded with the block
+                            continue
+
+                        # Check if we can put files in this block
+                        if not self.isBlockOpen(newFile = newFile,
+                                                block = currentBlock):
+                            # Then we have to close the block and get a new one
+                            currentBlock.setPendingAndCloseBlock()
+                            readyBlocks.append(currentBlock)
+                            currentBlock = self.getBlock(newFile = newFile,
+                                                         location = location)
+                            currentBlock.setAcquisitionEra(era = dspInfo['AcquisitionEra'])
+                            currentBlock.setProcessingVer(procVer = dspInfo['ProcessingVer'])
+
+                        # Now deal with the file
+                        currentBlock.addFile(newFile, self.datasetType, self.primaryDatasetType)
+                        self.filesToUpdate.append({'filelfn': newFile['lfn'],
+                                                   'block': currentBlock.getName()})
+                    # Done with the location
+                    readyBlocks.append(currentBlock)
+
         for block in readyBlocks:
             self.blockCache[block.getName()] = block
-        return
 
-    def checkTimeout(self):
-        """
-        _checkTimeout_
+        return
 
-        Loop all Open blocks and mark them as Pending if they have timed out.
-        """
-        for block in self.blockCache.values():
-            if block.status == "Open" and block.getTime() > block.getMaxBlockTime():
-                block.setPendingAndCloseBlock()
-                self.blockCache[block.getName()] = block
-    
-    def checkCompleted(self):
+    def checkBlockCompletion(self):
         """
-        _checkTimeout_
+        _checkBlockCompletion_
 
-        Loop all Open blocks and mark them as Pending if they have timed out.
+        Mark Open blocks as Pending if they have timed out or their workflows have completed
         """
         completedWorkflows = self.dbsUtil.getCompletedWorkflows()
         for block in self.blockCache.values():
             if block.status == "Open":
-                if block.workflow in completedWorkflows:
+                if ( block.getTime() > block.getMaxBlockTime() ) or any (key in completedWorkflows for key in block.workflows):
```

Comment:
I combined them because logically its a single condition. If (block either has waited over the max wait time or if any of the workflows producing files in the blocks has completed) then close block.

And why would I need to assign the block to the block cache ? Its already in there, I am iterating over the blocks in the block cache !